### PR TITLE
Microverse Project Enumerable Functions

### PR DESCRIPTION
**Pull request by Janluc**
## Enumerable Functions

- Custom versions of the standard Ruby Enumerable Functions
- Use of Yield as well as conditionals in order to receive desired output

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,46 @@
...
+
+## Show your support
+
+Give a ⭐️ if you like this project!
+
+## Acknowledgments
+
+- Hat tip to anyone whose code was used
+- Inspiration
+- etc
```
Comment (descholar-ceo) [inline README.md]: Please remove this `etc` it is not meaningful

File: README.md [hunk 2]
```diff
@@ -1 +1,46 @@
...
+
+## 📝 License
+
+This project is [MIT](lic.url) licensed.
```
Comment (descholar-ceo) [inline README.md]: This link is not working, fix it please or remove it

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
+
+module Enumerable
+  def my_each
+    counter = 0
+    while counter < size
+      if instance_of(Hash) || instance_of?(Range)
```
Comment (descholar-ceo) [inline main.rb]: The method is `instanceof?` (with question mark)

File: README.md [hunk 4]
```diff
@@ -1 +1,46 @@
...
+
+- Ruby,
+- Visual Studio Code, Git and Github.
+
+## Live Demo
+[Live Demo Link](https://repl.it/repls/ImmaculateTurbulentNumbers)
+
+## Authors
+
+👤 **Author1**
```
Comment (descholar-ceo) [inline README.md]: Add real names instead of `Author 1`

File: README.md [hunk 5]
```diff
@@ -1 +1,46 @@
...
+
+- Github: [@digitekh2i](https://https://github.com/digitekh2i)
+- Twitter: [@ObokoDaniel](https://twitter.com/ObokoDaniel)
+- Linkedin: [linkedin](http://linkedin.com/in/daniel-dikachi-1luvtek101)
+
+👤 **Author2**
```
Comment (descholar-ceo) [inline README.md]: Add real names instead of `Author 2`

File: README.md [hunk 6]
```diff
@@ -1 +1,46 @@
...
+- Github: [@Janluc](https://github.com/Janluc)   
+- Twitter: [@Janluc123](https://twitter.com/Janluc123)
+- Linkedin: [linkedin](https://www.linkedin.com/in/janluc-saneaux-91707a1b4/)
+
+
+## 🤝 Contributing
+
+Contributions, issues and feature requests are welcome!
+
+Feel free to check the [issues page](issues/).
```
Comment (descholar-ceo) [inline README.md]: Add a real issue link here, this one is not working

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        when Symbol
+          num = num.nil? ? item : num.send(accum, item)
+        when Numeric
+          accum = accum.send(sym, item)
+        end
+      end
+    end
+    num.nil? ? accum : num
+  end
+end
```
Comment (descholar-ceo) [inline main.rb:110]: #### my_select
- [ ] 7. #my_select returns an enumerator if no block is given

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    while counter < size
+      if instance_of(Hash) || instance_of?(Range)
+        new_hash = to_a
+        yield(new_hash[counter])
+      else
+        yield(self[counter])
+      end
+      counter += 1
+    end
+  end
```
Comment (descholar-ceo) [inline main.rb]: #### my_each
- [ ] 1. #my_each returns an Enumerator if no block is given
- [ ] 2. #my_each when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- [ ] 3. #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [ ] 4. #my_each_with_index returns an enumerator if no block is given
- [ ] 5. #my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- [ ] 6. #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  def my_all?(type = nil)
+    if block_given?
+      my_each { |item| return false if yield(item) == false }
+    elsif type.nil?
+      my_each { |item| return false if item.nil? }
+    else
+      my_each { |item| return false if item.class != type && item.class.superclass != type }
+    end
+    true
+  end
```
Comment (descholar-ceo) [inline main.rb:52]: #### my_all?
- [x] 8. #my_all? when no block or argument is given returns false when one of the collection members are false or nil
- [x] 9. #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] 10. #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: main.rb [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+  def my_any?(type = nil)
+    if block_given?
+      my_each { |item| return true if yield(item) == true }
+    elsif type.nil?
+      my_each { |item| return true if item.nil? }
+    else
+      my_each { |i| return true if i.instance_of?(type.class) || i.class.superclass == type || i.is_a?(type) }
+    end
+    false
+  end
```
Comment (descholar-ceo) [inline main.rb]: #### my_any?
- [x] 11. #my_any? when no block or argument is given returns false if at least one of the collection is not true
- [x] 12. #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
- [x] 13. #my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] 14. #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: main.rb [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  def my_none?(type = nil)
+    if block_given?
+      my_each { |item| return false if yield(item) == true }
+    elsif type.nil?
+      my_each { |item| return false if item.nil? && item != false }
+    else
+      my_each { |item| return false if item.instance_of?(type) || item.class.superclass == type || item.is_a?(type) }
+    end
+    true
+  end
```
Comment (descholar-ceo) [inline main.rb]: #### my_none?
- [x] 15. #my_none? when no block or argument is given returns true only if none of the collection members is true
- [x] 16. #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
- [x] 17. #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
- [x] 18. #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [x] 19. #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

File: main.rb [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+        case accum
+        when Symbol
+          num = num.nil? ? item : num.send(accum, item)
+        when Numeric
+          accum = accum.send(sym, item)
+        end
+      end
+    end
+    num.nil? ? accum : num
+  end
```
Comment (descholar-ceo) [inline main.rb:109]: #### my_inject
- [x] 25. #my_inject raises a "LocalJumpError" when no block or argument is given

File: main.rb [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+      my_each { |item| counter += 1 if item == count }
+    end
+    counter
+  end
+
+  def my_map
+    temp_arr = []
+    my_each { |item| temp_arr.push(yield(item)) unless yield(item) == false }
+    temp_arr
+  end
```
Comment (descholar-ceo) [inline main.rb:92]: #### my_map
- [x] 20. #my_map returns a new array with the results of running a given block once for every element in enum.
- [x] 21. #my_map returns a new array with the results of running a given block once for every element in enum::range
- [x] 22. #my_map returns an Enumerator if no block is given
- [x] 23. #my_map accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.
- [ ] 24. #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Change Required :vertical_traffic_light: :warning: 
## Hello @Janluc and @digitekh2i  :handshake: :handshake: :handshake: 
:+1: This is good to make it this far, your project is passing almost test-cases, congratulations on the good work you have done so far.

### :warning: However we need to make few changes to mark this project approved. See the requested change below :point_down: :point_down: :point_down: 
## Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes]
- [ ] Fix the `instance_of` method, I commented inline your codes, check them and fix it, it is causing your codes to fail running
- [ ] Your codes aren't passing some test cases, read below to understand what I mean:
```bash
# I tested your codes using the following values: 

  ARRAY_SIZE = 100
  LOWEST_VALUE = 0
  HIGHEST_VALUE = 9
  let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }
  let(:words) { %w[dog door rod blade] }
  let(:range) { Range.new(5, 50) }
  let(:hash) { { a: 1, b: 2, c: 3, d: 4, e: 5 } }
  let(:numbers) { [1, 2i, 3.14] }
  let!(:array_clone) { array.clone }

  # I ran 79 tests
  # => 25 test cases failed, see the failing cases in the comments i added in your codes
```
- [ ] Implement all test cases I added as inline comments in your codes, those are the failing test-cases

If anything seems to be unclear to you, don't hesitate to contact me for more clarification slack: `@Mugirase Emmanuel` | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
## Changes Requested #
- please go back and make the changes requested by the first tse
 do all of the changes requested for (my_each, my_all, my_any?, my_select?, my_none? , and my_map)

**Comment by Janluc** [comment]
> ## Changes Requested
> * please go back and make the changes requested by the first tse
>   do all of the changes requested for (my_each, my_all, my_any?, my_select?, my_none? , and my_map)

Hey thanks for the review! However, all the changes request by the TSE are provided in this code. The test cases to go along with the changes are also in the live demo link.

**Review by Georjane — APPROVED** [review]
# Status: PR Approved ✔️
## Hi @Janluc and @digitekh2i

Great job on your enumerable methods. 👏
They all have the same behavior like the normal enumerable methods. You can go ahead and merge this PR.

Goodluck on your next project 👍

**Pull request merged**